File: tools/trainer/hmmpak.py
# -*- coding: utf-8 -*-
"""HMMSYS PackFile (.pak / .bfhp) 讀寫器。

格式（逆向自 Celtic Kings: Rage of War，已對全部 6 個 pak 做過 byte-exact 往返驗證）::

    0x00  magic  b"HMMSYS PackFile\\n\\x1a"，補零到 0x20
    0x20  u32    fileCount
    0x24  u32    dirSize      目錄位元組數，從 0x28 起算
    0x28  目錄，每筆：
              u8   nameLen     完整檔名長度
              u8   prefixLen   與「前一筆檔名」共用的前綴長度
              u8   suffix[nameLen - prefixLen]
              u32  offset      payload 在檔案中的絕對位移
              u32  size        payload 位元組數
          u32  mtime[fileCount]    每檔一筆 DOS 日期時間
          payload 區，依目錄順序連續排列，無對齊、無間隙

檔名一律大寫、以反斜線分隔，對應遊戲內的 ``data/...`` 虛擬路徑。
"""

import struct

MAGIC = b'HMMSYS PackFile\n\x1a'
HEADER_SIZE = 0x28
DEFAULT_MTIME = 0x303765D0  # 2004-01-23 12:46:32，原版 data.pak 的時間戳


class PakError(Exception):
    pass


class Entry(object):
    __slots__ = ('name', 'offset', 'size', 'mtime')

    def __init__(self, name, offset, size, mtime=DEFAULT_MTIME):
        self.name = name
        self.offset = offset
        self.size = size
        self.mtime = mtime

    def __repr__(self):
        return '<Entry %s off=0x%X size=%d>' % (self.name, self.offset, self.size)


def _common_prefix(a, b):
    n = min(len(a), len(b), 255)
    i = 0
    while i < n and a[i] == b[i]:
        i += 1
    return i


def read_directory(data):
    """解析目錄，回傳 list[Entry]。"""
    if data[:len(MAGIC)] != MAGIC:
        raise PakError('不是 HMMSYS PackFile（magic 不符：%r）' % data[:16])

    count, dirsize = struct.unpack_from('<II', data, 0x20)
    dir_end = HEADER_SIZE + dirsize
    if dir_end > len(data):
        raise PakError('目錄長度超出檔案大小')

    entries = []
    pos = HEADER_SIZE
    prev = ''
    for _ in range(count):
        name_len = data[pos]
        prefix_len = data[pos + 1]
        if prefix_len > name_len:
            raise PakError('目錄損毀：prefixLen > nameLen @0x%X' % pos)
        suffix_end = pos + 2 + (name_len - prefix_len)
        name = prev[:prefix_len] + data[pos + 2:suffix_end].decode('latin-1')
        offset, size = struct.unpack_from('<II', data, suffix_end)
        if offset + size > len(data):
            raise PakError('項目 %s 的資料超出檔案範圍' % name)
        entries.append(Entry(name, offset, size))
        prev = name
        pos = suffix_end + 8

    if pos != dir_end:
        raise PakError('目錄尾端不對齊：0x%X != 0x%X' % (pos, dir_end))

    mtimes = struct.unpack_from('<%dI' % count, data, dir_end)
    for entry, mtime in zip(entries, mtimes):
        entry.mtime = mtime
    return entries


class Pak(object):
    """一個載入到記憶體的 pak。名稱查詢不分大小寫。"""

    def __init__(self, entries, blobs):
        self.entries = entries          # list[Entry]，保持原始順序
        self._blobs = blobs             # dict: 正規化名稱 -> bytes
        self._index = dict((e.name.upper(), e) for e in entries)

    # ---- 建立 ----------------------------------------------------------
    @classmethod
    def from_bytes(cls, data):
        entries = read_directory(data)
        blobs = {}
        for e in entries:
            blobs[e.name.upper()] = data[e.offset:e.offset + e.size]
        return cls(entries, blobs)

    @classmethod
    def load(cls, path):
        with open(path, 'rb') as f:
            return cls.from_bytes(f.read())

    # ---- 讀取 ----------------------------------------------------------
    def __contains__(self, name):
        return name.upper() in self._blobs

    def __len__(self):
        return len(self.entries)

    def names(self):
        return [e.name for e in self.entries]

    def read(self, name):
        key = name.upper()
        if key not in self._blobs:
            raise KeyError('pak 內沒有 %s' % name)
        return self._blobs[key]

    def read_text(self, name, encoding='latin-1'):
        return self.read(name).decode(encoding)

    # ---- 寫入 ----------------------------------------------------------
    def write(self, name, blob):
        """覆寫既有檔案，或新增一筆（新增時排在目錄最後）。"""
        if not isinstance(blob, bytes):
            raise TypeError('blob 必須是 bytes')
        key = name.upper()
        if key in self._blobs:
            self._blobs[key] = blob
            self._index[key].size = len(blob)
        else:
            entry = Entry(key, 0, len(blob))
            self.entries.append(entry)
            self._index[key] = entry
            self._blobs[key] = blob

    def write_text(self, name, text, encoding='latin-1'):
        self.write(name, text.encode(encoding))

    # ---- 序列化 --------------------------------------------------------
    def to_bytes(self):
        names = [e.name for e in self.entries]

        dirsize = 0
        prev = ''
        for name in names:
            dirsize += 2 + (len(name) - _common_prefix(prev, name)) + 8
            prev = name

        data_start = HEADER_SIZE + dirsize + 4 * len(names)

        out = bytearray(MAGIC)
        out += b'\x00' * (0x20 - len(MAGIC))
        out += struct.pack('<II', len(names), dirsize)

        offset = data_start
        prev = ''
        for entry in self.entries:
            name = entry.name
            prefix_len = _common_prefix(prev, name)
            blob = self._blobs[name.upper()]
            out += bytes((len(name), prefix_len))
            out += name[prefix_len:].encode('latin-1')
            out += struct.pack('<II', offset, len(blob))
            offset += len(blob)
            prev = name

        out += struct.pack('<%dI' % len(names), *[e.mtime for e in self.entries])
        assert len(out) == data_start, (len(out), data_start)

        for entry in self.entries:
            out += self._blobs[entry.name.upper()]
        return bytes(out)

    def save(self, path):
        with open(path, 'wb') as f:
            f.write(self.to_bytes())

File: src/CKToolkit/Core/Common/HmmPak.cs
using System.Text;

namespace CKToolkit.Core.Common;

/// <summary>
/// HMMSYS PackFile（《Celtic Kings》專用之 .pak / .bfhp 封裝檔）讀寫器。
///
/// 格式規格：
///   0x00  magic "HMMSYS PackFile\n\x1a"，補零到 0x20
///   0x20  u32 fileCount
///   0x24  u32 dirSize            目錄位元組數，自 0x28 起算
///   0x28  目錄區，每筆條目：
///           u8  nameLen          完整檔名長度
///           u8  prefixLen        與前一筆檔名共用的前綴長度
///           u8  suffix[nameLen - prefixLen]
///           u32 offset           payload 絕對檔案位移
///           u32 size             payload 位元組長度
///         u32 mtime[fileCount]   每檔一筆 DOS 日期時間戳
///         payload 資料區，依目錄順序連續排列，無對齊、無間隙
///
/// 起始實作來自修改器之 HmmPak.cs（已對遊戲全部 6 個 HMMSYS pak 含 136 MB assets.pak
/// 通過逐位元組往返驗證）。
/// </summary>
public sealed class HmmPak
{
    /// <summary>pak 內的檔名一律以 latin-1 存放。</summary>
    public static readonly Encoding PakEncoding = Encoding.Latin1;

    private static ReadOnlySpan<byte> Magic => "HMMSYS PackFile\n\x1a"u8;

    public const int HeaderSize = 0x28;

    /// <summary>原版 data.pak 多數項目的時間戳：2004-01-23 12:46:32。</summary>
    public const uint DefaultMTime = 0x303765D0;

    private readonly List<PakEntry> _entries;
    private readonly Dictionary<string, byte[]> _blobs;

    private HmmPak(List<PakEntry> entries, Dictionary<string, byte[]> blobs)
    {
        _entries = entries;
        _blobs = blobs;
    }

    public static HmmPak CreateEmpty() =>
        new([], new Dictionary<string, byte[]>(StringComparer.OrdinalIgnoreCase));

    public IReadOnlyList<PakEntry> Entries => _entries;

    public int Count => _entries.Count;

    public static string Normalize(string name) =>
        name.Replace('/', '\\').ToUpperInvariant();

    // ---- 讀取與還原 --------------------------------------------------------

    public static HmmPak FromBytes(byte[] data)
    {
        if (data.Length < HeaderSize || !data.AsSpan(0, Magic.Length).SequenceEqual(Magic))
            throw new PakException("不是 HMMSYS PackFile：magic 不符");

        int count = (int)BitConverter.ToUInt32(data, 0x20);
        int dirSize = (int)BitConverter.ToUInt32(data, 0x24);
        int dirEnd = HeaderSize + dirSize;
        if (count < 0 || dirSize < 0 || dirEnd > data.Length)
            throw new PakException("目錄長度超出檔案大小");

        var entries = new List<PakEntry>(count);
        var blobs = new Dictionary<string, byte[]>(count, StringComparer.OrdinalIgnoreCase);

        int pos = HeaderSize;
        string prev = string.Empty;
        for (int i = 0; i < count; i++)
        {
            if (pos + 2 > dirEnd) throw new PakException("目錄提前結束");
            int nameLen = data[pos];
            int prefixLen = data[pos + 1];
            if (prefixLen > nameLen || prefixLen > prev.Length)
                throw new PakException($"目錄損毀：第 {i} 筆的前綴長度不合理");

            int suffixLen = nameLen - prefixLen;
            int suffixEnd = pos + 2 + suffixLen;
            if (suffixEnd + 8 > dirEnd) throw new PakException("目錄提前結束");

            string name = string.Concat(
                prev.AsSpan(0, prefixLen),
                PakEncoding.GetString(data, pos + 2, suffixLen));

            uint offset = BitConverter.ToUInt32(data, suffixEnd);
            uint size = BitConverter.ToUInt32(data, suffixEnd + 4);
            if (offset + (long)size > data.Length)
                throw new PakException($"項目 {name} 的資料超出檔案範圍");

            entries.Add(new PakEntry(name, size));
            blobs[name] = data.AsSpan((int)offset, (int)size).ToArray();

            prev = name;
            pos = suffixEnd + 8;
        }

        if (pos != dirEnd)
            throw new PakException($"目錄尾端不對齊：0x{pos:X} != 0x{dirEnd:X}");

        for (int i = 0; i < count; i++)
            entries[i].MTime = BitConverter.ToUInt32(data, dirEnd + i * 4);

        return new HmmPak(entries, blobs);
    }

    public static HmmPak Load(string path) => FromBytes(File.ReadAllBytes(path));

    public bool Contains(string name) => _blobs.ContainsKey(Normalize(name));

    public int IndexOf(string name)
    {
        string key = Normalize(name);
        for (int i = 0; i < _entries.Count; i++)
        {
            if (string.Equals(_entries[i].Name, key, StringComparison.OrdinalIgnoreCase))
                return i;
        }
        return -1;
    }

    public byte[] Read(string name)
    {
        string key = Normalize(name);
        return _blobs.TryGetValue(key, out var blob)
            ? blob
            : throw new PakException($"pak 內沒有檔案：{name}");
    }

    public string ReadText(string name, Encoding? encoding = null) =>
        (encoding ?? PakEncoding).GetString(Read(name));

    public IEnumerable<string> Names() => _entries.Select(e => e.Name);

    // ---- 寫入與編輯 --------------------------------------------------------

    /// <summary>覆寫既有檔案；不存在則新增一筆（排在目錄最後）。</summary>
    public void Write(string name, byte[] blob, uint? mtime = null)
    {
        string key = Normalize(name);
        if (_blobs.ContainsKey(key))
        {
            _blobs[key] = blob;
            foreach (var e in _entries)
            {
                if (string.Equals(e.Name, key, StringComparison.OrdinalIgnoreCase))
                {
                    e.Size = (uint)blob.Length;
                    if (mtime.HasValue) e.MTime = mtime.Value;
                    break;
                }
            }
        }
        else
        {
            var entry = new PakEntry(key, (uint)blob.Length);
            if (mtime.HasValue) entry.MTime = mtime.Value;
            _entries.Add(entry);
            _blobs[key] = blob;
        }
    }

    public void WriteText(string name, string text, Encoding? encoding = null, uint? mtime = null) =>
        Write(name, (encoding ?? PakEncoding).GetBytes(text), mtime);

    public bool Remove(string name)
    {
        string key = Normalize(name);
        if (!_blobs.Remove(key)) return false;
        _entries.RemoveAll(e => string.Equals(e.Name, key, StringComparison.OrdinalIgnoreCase));
        return true;
    }

    // ---- 序列化 ------------------------------------------------------------

    public byte[] ToBytes()
    {
        // ★ 目錄必須依名稱排序後才序列化。
        //
        // HMMSYS 的目錄是排序的（原版 local.pak 924 個項目、data.pak 876 個項目都完全有序），
        // 而且採前綴壓縮——每個項目只存「與前一個名稱共用幾個字元」。引擎顯然依賴這個
        // 順序查表（極可能是二分搜尋）。
        //
        // 這條規則違反時的症狀非常有欺騙性：檔案內容完全正確、雜湊也對，但遊戲找不到它們。
        // 實際發生過：語言包安裝後 CHINESE\ 底下 297 個項目全部存在且內容正確，
        // vxSettings 也指向 chinese，但遊戲仍顯示英文——因為新項目被 append 在目錄尾端，
        // 排在 SCENARIOS\... 之後，查表全部落空。前身 Python 實作在 ckpatch.py:308
        // 用 sorted(files.items()) 排序後才建檔，所以它沒有這個問題。
        //
        // 排序規則：名稱已由 Normalize 統一為大寫加反斜線，用序數比較即與 Python 的
        // 預設字串排序一致。
        _entries.Sort(static (a, b) => string.CompareOrdinal(a.Name, b.Name));

        int dirSize = 0;
        string prev = string.Empty;
        foreach (var e in _entries)
        {
            dirSize += 2 + (e.Name.Length - CommonPrefix(prev, e.Name)) + 8;
            prev = e.Name;
        }

        int dataStart = HeaderSize + dirSize + 4 * _entries.Count;
        long total = dataStart;
        foreach (var e in _entries) total += _blobs[e.Name].Length;

        var output = new byte[total];
        var span = output.AsSpan();

        Magic.CopyTo(span);
        BitConverter.TryWriteBytes(span[0x20..], (uint)_entries.Count);
        BitConverter.TryWriteBytes(span[0x24..], (uint)dirSize);

        int pos = HeaderSize;
        uint offset = (uint)dataStart;
        prev = string.Empty;
        foreach (var e in _entries)
        {
            int prefixLen = CommonPrefix(prev, e.Name);
            var blob = _blobs[e.Name];

            output[pos++] = (byte)e.Name.Length;
            output[pos++] = (byte)prefixLen;
            pos += PakEncoding.GetBytes(e.Name, prefixLen, e.Name.Length - prefixLen, output, pos);

            BitConverter.TryWriteBytes(span[pos..], offset);
            BitConverter.TryWriteBytes(span[(pos + 4)..], (uint)blob.Length);
            pos += 8;

            offset += (uint)blob.Length;
            prev = e.Name;
        }

        foreach (var e in _entries)
        {
            BitConverter.TryWriteBytes(span[pos..], e.MTime);
            pos += 4;
        }

        if (pos != dataStart)
            throw new PakException($"目錄寫入長度不符：0x{pos:X} != 0x{dataStart:X}");

        foreach (var e in _entries)
        {
            var blob = _blobs[e.Name];
            blob.CopyTo(span[pos..]);
            pos += blob.Length;
        }

        return output;
    }

    public void Save(string path) => File.WriteAllBytes(path, ToBytes());

    private static int CommonPrefix(string a, string b)
    {
        int max = Math.Min(Math.Min(a.Length, b.Length), 255);
        int i = 0;
        while (i < max && a[i] == b[i]) i++;
        return i;
    }
}

public sealed class PakEntry(string name, uint size)
{
    public string Name { get; } = name;
    public uint Size { get; internal set; } = size;
    public uint MTime { get; internal set; } = HmmPak.DefaultMTime;

    public override string ToString() => $"{Name} ({Size} bytes)";
}

public sealed class PakException(string message) : Exception(message);
